clustered/server/src/main/java/org/ehcache/clustered/server/store/ClusterTierActiveEntity.java

```diff
@@ -586,40 +594,74 @@ public void synchronizeKeyToPassive(PassiveSynchronizationChannel<EhcacheEntityM
-          } catch (TimeoutException e) {
-            throw new AssertionError("Server side store is not expected to throw timeout exception");
-          }
-          for (Element element : chain) {
-            size.addAndGet(element.getPayload().remaining());
-          }
-          mappingsToSend.get().put(key, chain);
-          if (size.get() > dataSizeThreshold) {
-            syncChannel.synchronizeToPassive(new EhcacheDataSyncMessage(mappingsToSend.get()));
-            mappingsToSend.set(new HashMap<>());
-            size.set(0);
+      CompletableFuture<Void> asyncGet = CompletableFuture.runAsync(() -> doGetsForSync(segmentId, messageQ));
+      try {
+        boolean done = false;
+        while (!done) {
+          EhcacheDataSyncMessage msg = messageQ.take();
+          if (msg instanceof LastDataSyncMessage) {
+            done = true;
+          } else {
+            syncChannel.synchronizeToPassive(msg);
           }
-        });
-      if (!mappingsToSend.get().isEmpty()) {
-        syncChannel.synchronizeToPassive(new EhcacheDataSyncMessage(mappingsToSend.get()));
-        mappingsToSend.set(new HashMap<>());
-        size.set(0);
+        }
+        asyncGet.get();
+      } catch (InterruptedException | ExecutionException e) {
+        throw new RuntimeException(e);
       }
     }
     sendMessageTrackerReplication(syncChannel, concurrencyKey - 1);
 
     LOGGER.info("Sync complete for concurrency key {}.", concurrencyKey);
   }
 
+  private void doGetsForSync(int segmentId, BlockingQueue<EhcacheDataSyncMessage> messageQ) {
+    if (dataSizeThreshold == null) {
+      // do it only once
+      dataSizeThreshold = Long.getLong(SYNC_DATA_SIZE_PROP, DEFAULT_SYNC_DATA_SIZE_THRESHOLD);
+      dataGetsThreshold = Integer.getInteger(SYNC_DATA_GETS_PROP, DEFAULT_SYNC_DATA_GETS_THRESHOLD);
+    }
+    ServerSideServerStore store = stateService.getStore(storeIdentifier);
+    int numGets = 0;
+    long size = 0;
+    Map<Long, Chain> mappingsToSend = new HashMap<>();
+    try {
+      for (Long key : store.getSegmentKeySets().get(segmentId)) {
+        final Chain chain;
+        try {
+          chain = store.get(key);
+          numGets++;
+        } catch (TimeoutException e) {
+          throw new AssertionError("Server side store is not expected to throw timeout exception");
+        }
+        for (Element element : chain) {
+          size += element.getPayload().remaining();
+        }
+        mappingsToSend.put(key, chain);
+        if (size > dataSizeThreshold || numGets > dataGetsThreshold) {
+          messageQ.put(new EhcacheDataSyncMessage(mappingsToSend));
+          mappingsToSend = new HashMap<>();
+          size = 0;
+          numGets = 0;
+        }
+      }
+      if (!mappingsToSend.isEmpty()) {
+        messageQ.put(new EhcacheDataSyncMessage(mappingsToSend));
+      }
+      messageQ.put(new LastDataSyncMessage());
+    } catch (InterruptedException e) {
+      throw new RuntimeException(e);
```

Comment:
From my point of view, the exception will probably be caught by the common pool and disappear. The interruption should be handled to get out nicely. And throwables should be logged and rethrown I guess.